### DISCUSSION TITLE
添加新翻译源的集中反馈区

### DISCUSSION THREAD

**Discussion by lmk123** [hcfyapp/crx-selection-translate#930] [讨论]
鉴于有很多人都希望划词翻译能添加更多翻译源，所以我单独开了这个帖子用于集中收集大家关于这方面的建议。

从目前的反馈来看，以下翻译源是大家希望添加的：

- DeepL https://www.deepl.com/translator +18 （Done）
- 阿里翻译 https://translate.alibaba.com/ +4 （Done）
- 腾讯翻译君 https://fanyi.qq.com/  +4 （Done）
- 汉典 http://www.zdic.net/ +3
- 剑桥词典 https://dictionary.cambridge.org/zhs/ +2
- Yandex https://translate.yandex.com/ +1
- 搜狗翻译 https://fanyi.sogou.com/ +2 （Done）
- 金山词霸（爱词霸）http://www.iciba.com/ +3
- 彩云小译 https://fanyi.caiyunapp.com/ +2 （Done）
- Online Etymology Dictionary https://www.etymonline.com/ +1
- Dictionary.com https://www.dictionary.com/ +1
- Vocabulary.com Dictionary https://www.vocabulary.com/dictionary/ +1
- YourDictionary https://www.yourdictionary.com/ +1

如果你觉得这里面有你想要添加的，请在下面回复，例如 “DeepL +1”；如果你有别的翻译源想要推荐也请留言告知。留言后最好能提供你希望加入这个翻译源的原因，例如“这个翻译某某语言比较准”、“英文翻译方面比 xxx 好”之类的。

我会根据大家的反馈将一些翻译源优先加入划词翻译，目前我的计划是：

- 划词翻译会提供以上大部分翻译服务的跳转链接，点击之后就会打开一个新标签页，显示它们各自的翻译结果网页。目前计划加入 阿里翻译、腾讯翻译君、搜狗翻译等外部链接。
- 对于提供官方接口的翻译服务提供直接显示翻译结果的方式，但需要自行申请 API key 之后填写进划词翻译。目前计划加入的有百度、有道、腾讯和阿里，它们都有免费使用额度。
- 对于部分翻译服务，我会考虑直接显示翻译结果，目前计划加入 DeepL、汉典、剑桥词典。

## 进展

- v7.4.3 已添加了 DeepL（直接显示）和阿里翻译、腾讯翻译君、搜狗翻译、金山词霸、有道词典的外部跳转链接。
- v7.5.12 添加了阿里翻译（需申请 API Key）
- v7.5.13 添加了腾讯翻译君（需申请 API Key）
- v7.5.14 添加了搜狗翻译和彩云小译（都需要申请 APi Key）

**Comment by butaixianran** [comment]
彩云小译 +1
彩云小译日文翻译比百度更准确。原因参考：[#466](https://github.com/lmk123/crx-selection-translate/issues/466)

**Comment by butaixianran** [comment]
**关于DeepL：各位，DeepL并不像大家想的那么好。**

**DeepL的问题：**
**DeepL实际上，和Google翻译一样，有个根本问题：“传话筒接力”。**

Google翻译不准确的一个巨大问题，在于虽然Google也是AI翻译，但是，他把所有语言翻译为英文，然后，再把这个生成的英文，翻译成用户选择的目标语言。

这个2次翻译，就会导致巨大问题，因为翻译为中间语言的词汇，可能是多义词。于是，把这个多义词2次翻译为目标语言，往往会选择错误的意思。

比如，日语“処世術”，用户选择翻译为中文。Google会先翻译为英文"treatment"。这个词，有“对待方法”和“治疗”2个意思。再翻译为中文时，会选择“治疗”，于是Google就翻译错了。

deepl也是类似，他把所有语言，先翻译成德语，再翻译成用户选择的目标语言。它比google准确在于，deepl以前是做电子词典的，所以AI的词汇量比google大，德语词汇歧义比英语少。但是，并没有解决2次翻译导致的“传话筒接力”这个根本问题。

本人也经常使用deepl，但很快发现，翻译错误、甚至完全无法翻译的现象非常多。就是因为“传话筒接力”问题。

而国内翻译工具，都是AI翻译，引擎和deepl一样，而且没有“传话筒接力”问题，因此准确度实际上比deepl要好，仅仅只是没有宣传，缺少名气。

**本人对比过国内AI翻译引擎，其中，行文最通畅，最像人工翻译的引擎，实际上是阿里云的翻译服务！**

行文流程程度比deepl更好，词汇量和腾讯翻译一致。但是，阿里云的翻译从未做过宣传，因此大众不知道。作为质量最高的翻译服务，这里的引擎列表里甚至没有。

因此，希望降低deepl的优先级。大众用户是非常容易被名气误导的，因为从未对比，根本不知道其他选择。

如果要做，建议所有收费翻译服务，做成让用户填入自己的key或token的方式使用。谁用谁去申请。

**最后：**
**阿里云翻译 +1**

**大家可以去体会下阿里云翻译：**
[https://www.aliyun.com/product/ai/base_alimt](https://www.aliyun.com/product/ai/base_alimt)

更新：
阿里云API也是收费的，但是每月有百万字符的免费额度。做成填入用户自己的api key的方式，可以比较好的解决费用问题。

**Comment by linexu** [comment]
DeepL +++N（請問作者不加PRO，只加普通免費版可以嗎？）
腾讯翻译君 -N
汉典 -N
剑桥词典 -N
Yandex -N
搜狗翻译 -N
金山词霸（爱词霸） -N
彩云小译 -N
Online Etymology Dictionary -N
Dictionary.com -N
Vocabulary.com Dictionary  -N
YourDictionary  -N

**Comment by lmk123** [comment]
@linexu 
> DeepL +++N（請問作者不加PRO，只加普通免費版可以嗎？）

我试过了，目前在技术上是有难度的，所以有可能会用收费接口。如果大家接受不了收费接口，那么我就不加入了。

**Comment by linexu** [comment]
> @linexu
> 
> > DeepL +++N（請問作者不加PRO，只加普通免費版可以嗎？）
> 
> 我试过了，目前在技术上是有难度的，所以有可能会用收费接口。如果大家接受不了收费接口，那么我就不加入了。

在Tampermonkey裡找到一個插件，名稱為“Translate” by https://github.com/barrer
看他的代碼裡面加入了 name: 'DeepL翻译', ，感覺應該不難。此插件也加入了太多的翻譯源。

另外提一句，目前的“谷歌翻譯”完全就是顛倒黑白、指鹿為馬。簡直就是魔鬼作惡。（太多的例證）

**Comment by butaixianran** [comment]
> 在Tampermonkey裡找到一個插件，名稱為“Translate” by https://github.com/barrer
> 看他的代碼裡面加入了 name: 'DeepL翻译', ，感覺應該不難。此插件也加入了太多的翻譯源。
> 

这个作者没有靠API接入deepl，他的脚本是生成一个跳转到deepl的链接，链接包含了要翻译的文字。这样就避免了API费用，而且脚本开发也非常简单，几行代码就能添加一个引擎。

**Comment by lmk123** [comment]
如果只是添加一个跳转链接的话，那倒是很简单，我可以在下个版本中先将这些翻译源用外部链接的方式接入进来。

**Comment by DesistDaydream** [comment]
实在抱歉，没有仔细看 ISSUE 就单独提了。

@butaixianran 非常感谢，测试了一下阿里云的翻译，感觉也非常不错，与 DeepL 各有优劣(速度较慢是通病)，有阿里云坐镇，还是国内产品，也不怕被墙。

DeepL +1
阿里云翻译 +1

> **关于DeepL：各位，DeepL并不像大家想的那么好。**
> 
> **DeepL的问题：**
> **DeepL实际上，和Google翻译一样，有个根本问题：“传话筒接力”。**
> 
> Google翻译不准确的一个巨大问题，在于虽然Google也是AI翻译，但是，他把所有语言翻译为英文，然后，再把这个生成的英文，翻译成用户选择的目标语言。
> 
> 这个2次翻译，就会导致巨大问题，因为翻译为中间语言的词汇，可能是多义词。于是，把这个多义词2次翻译为目标语言，往往会选择错误的意思。
> 
> 比如，日语“処世術”，用户选择翻译为中文。Google会先翻译为英文"treatment"。这个词，有“对待方法”和“治疗”2个意思。再翻译为中文时，会选择“治疗”，于是Google就翻译错了。
> 
> deepl也是类似，他把所有语言，先翻译成德语，再翻译成用户选择的目标语言。它比google准确在于，deepl以前是做电子词典的，所以AI的词汇量比google大，德语词汇歧义比英语少。但是，并没有解决2次翻译导致的“传话筒接力”这个根本问题。
> 
> 本人也经常使用deepl，但很快发现，翻译错误、甚至完全无法翻译的现象非常多。就是因为“传话筒接力”问题。
> 
> 而国内翻译工具，都是AI翻译，引擎和deepl一样，而且没有“传话筒接力”问题，因此准确度实际上比deepl要好，仅仅只是没有宣传，缺少名气。
> 
> **本人对比过国内AI翻译引擎，其中，行文最通畅，最像人工翻译的引擎，实际上是阿里云的翻译服务！**
> 
> 行文流程程度比deepl更好，词汇量和腾讯翻译一致。但是，阿里云的翻译从未做过宣传，因此大众不知道。作为质量最高的翻译服务，这里的引擎列表里甚至没有。
> 
> 因此，希望降低deepl的优先级。大众用户是非常容易被名气误导的，因为从未对比，根本不知道其他选择。
> 
> 如果要做，建议所有收费翻译服务，做成让用户填入自己的key或token的方式使用。谁用谁去申请。
> 
> **最后：**
> **阿里云翻译 +1**
> 
> **大家可以去体会下阿里云翻译：**
> https://www.aliyun.com/product/ai/base_alimt
> 
> 更新：
> 阿里云API也是收费的，但是每月有百万字符的免费额度。做成填入用户自己的api key的方式，可以比较好的解决费用问题。

**Comment by butaixianran** [comment]
在作者没有添加其他引擎之前，可以先用下面这个脚本实现多引擎：
https://github.com/butaixianran/AI-Translate-and-Search

涵盖了以下功能：
**划词AI翻译+搜索：**  
* **AI翻译按质量和热度综合排序。调用：** 金山翻译，阿里翻译，百度翻译，搜狗翻译，腾讯翻译，彩云小译，DeepL，沪江日语词典等。
* **搜索包含：** 知乎，b站，微博，百科, wiki，youtube，twitter等。
* **设置：** 展开后，点击“Setting”可以调整 排序 和 是否显示。


**实现机制说明：**  
选择文字之后，会出现图标，点击图标，会跳到对应的翻译服务页面。脚本会自动填入需要翻译的文本，模拟按键事件，触发翻译。这样就避免了API费用，还迅速享受了各个翻译服务。  

而且，因为是跳转，也能实现划词搜索。

做 划词跳转 的脚本非常多，但是能解析翻译页面，自动填入文本，自动触发翻译事件的，目前就这一个。

**Comment by DesistDaydream** [comment]
> 在作者没有添加其他引擎之前，可以先用下面这个脚本实现多引擎：
> https://github.com/butaixianran/AI-Translate-and-Search
> 
> 涵盖了以下功能：
> **划词AI翻译+搜索：**
> 
> * **AI翻译按质量和热度综合排序。调用：** 金山翻译，阿里翻译，百度翻译，搜狗翻译，腾讯翻译，彩云小译，DeepL，沪江日语词典等。
> * **搜索包含：** 知乎，b站，微博，百科, wiki，youtube，twitter等。
> * **设置：** 展开后，点击“Setting”可以调整 排序 和 是否显示。
> 
> **实现机制说明：**
> 选择文字之后，会出现图标，点击图标，会跳到对应的翻译服务页面。脚本会自动填入需要翻译的文本，模拟按键事件，触发翻译。这样就避免了API费用，还迅速享受了各个翻译服务。
> 
> 而且，因为是跳转，也能实现划词搜索。
> 
> 做 划词跳转 的脚本非常多，但是能解析翻译页面，自动填入文本，自动触发翻译事件的，目前就这一个。

功能相当好用，居然还集合了各种搜索引擎，而且开启关闭各种选项都相当方便。只不过开了这个，就没法用本来的划词翻译了。。。这就会导致没法在本页面显示结果。。。。这个体验有点别扭。。。

**Comment by lmk123** [comment]
@DesistDaydream @butaixianran 我会做两手准备，比如像百度、有道、腾讯、阿里翻译这种，我会既提供外部跳转链接，也提供直接显示翻译结果的方式，但是直接显示翻译结果需要你自行去它们的官网申请 API Key，当然我也会提供详细的申请步骤文档。

目前大部分官方接口都提供每月免费额度，所以我觉得这应该是既解决了收费问题、又提供了最大程度的用户体验的方式。

**Comment by butaixianran** [comment]
> 功能相当好用，居然还集合了各种搜索引擎，而且开启关闭各种选项都相当方便。只不过开了这个，就没法用本来的划词翻译了。。。这就会导致没法在本页面显示结果。。。。这个体验有点别扭。。。

非要在当前页面显示的话，可以结合 **百度翻译的客户端** 或 **金山翻译的客户端**。金山AI翻译的质量参考这个脚本的github说明页面。

**Comment by lmk123** [comment]
v7.5.12 已加入阿里云翻译。

**Comment by lmk123** [comment]
v7.5.13 已加入腾讯翻译君。

**Comment by lmk123** [comment]
v7.5.14 添加了搜狗翻译和彩云小译

**Comment by GH01** [comment]
《Bing词典》只支持英文吗？比如日文/俄文就报错。
而且只是词典，不支持长文。

《Bing翻译》没有接口吗？
貌似它不支持链接形式……

**Comment by lmk123** [comment]
是的，必应词典只支持中文和英文，实际上就跟你在 https://cn.bing.com/dict/ 搜索是一样的。

必应翻译也有接口，目前还没集成进来，不过你可以用快捷链接的形式使用，如 https://www.bing.com/translator/?ref=TThis&text=test&from=en&to=zh-Hans

**Comment by HarryHuaa** [comment]
vocabulary.com +1
urban.com+1
etymonline.com+1
